### ISSUE TITLE
API.DiffTreeVertexDiffTerm should not be a sum type.

### ISSUE THREAD

**Issue by patrickt** [github/semantic#168]
The current definition is an uncomfortable hybrid of a message (a record type) and a field (a sum type):

```haskell
data DiffTreeVertexDiffTerm
  = Deleted (Maybe DeletedTerm)
  | Inserted (Maybe InsertedTerm)
  | Replaced (Maybe ReplacedTerm)
  | Merged (Maybe MergedTerm)
  deriving stock (Eq, Ord, Show, Generic)
  deriving anyclass (Proto3.Message, Proto3.Named, NFData)
```

To abide by protobuf standards, this should be closer to

```haskell
data DiffTreeVertexDiffTerm = DiffTreeVertexDiffTerm
  { deleted :: Maybe DeletedTerm
  , inserted :: Maybe InsertedTerm
  , replaced :: Maybe ReplacedTerm
  , merged :: Maybe MergedTerm
  }
```

which is much closer to the actual protobuf definition:

```proto
  oneof diff_term {
    DeletedTerm deleted = 2;
    InsertedTerm inserted = 3;
    ReplacedTerm replaced = 4;
    MergedTerm merged = 5;
  }
```

**Comment by patrickt** [comment]
This is blocking #111.

**Comment by robrix** [comment]
The product type has strictly less information than the sum type does. Is that gonna be ok?

**Comment by patrickt** [comment]
Yes—indeed, we need it to hold less information for forward-compatibility. The case of `DiffTreeVertexDiffTerm Nothing Nothing Nothing Nothing` seems pathological but is entirely possible if we added a new alternative to the `oneof` within `diff_term`—consumers that hadn’t updated to see that new field should yield a term full of `Nothing`s. We can yield a nice API, however, with `-XPatternSynonyms`:

```
pattern Deleted :: DeletedTerm -> DiffTreeVertexDiffTerm
pattern Deleted a = DiffTreeVertexDiffTerm (Just a) Nothing Nothing Nothing
```

**Comment by robrix** [comment]
Gotcha. Thanks v. much @patrickt!

**Comment by robrix** [comment]
Reopened by #205.

**Comment by patrickt** [comment]
Fixed by #296.
